### ISSUE TITLE
Why are you increasing the number of instance masks in the final prediction step?

### ISSUE THREAD

**Issue by aminebdj** [VinAIResearch/Open3DIS#17]
Hello,

I hope you are doing well,

I just noticed that the number of instances in the final step is being increased in the final step of prediction [here](https://github.com/VinAIResearch/Open3DIS/blob/1e5ac0eb6ec75e44abfef2d9c970f05cf2c63ad3/tools/generate_3d_inst.py#L169). Since the prediction are forced to have confidence of 1.0, the final evaluation score will be sensitive, and adding additional duplicate masks would increase the score, which makes the results less credible. I tried removing the duplicate masks from the masks you provided here [here](https://drive.google.com/file/d/1DpuNSdLjNHIjjNr6fAu6oVwVRabatnpy/view?usp=sharing) and the mAP drops from 23.0 to 16.8. I also evaluated the method with ground truth masks and duplicates will result in an increase of 10% from 30 to 40 (I assume because the evaluation is sensitive with only two confidences of 0 and 1). Would it be possible to confirm whether the results can be reproduced without using duplicate masks as in [here](https://github.com/VinAIResearch/Open3DIS/blob/1e5ac0eb6ec75e44abfef2d9c970f05cf2c63ad3/tools/generate_3d_inst.py#L169)? 

Here is the code on how I am removing the duplicates, I added it after this line [here](https://github.com/VinAIResearch/Open3DIS/blob/1e5ac0eb6ec75e44abfef2d9c970f05cf2c63ad3/tools/generate_3d_inst.py#L286):

```
masks_final_, inverse_map = masks_final.unique(dim=0, return_inverse = True)
inverse_map = inverse_map.unique()

masks_final = masks_final[inverse_map]
scores_final = scores_final[inverse_map]
cls_final = cls_final[inverse_map]
```

many thanks,

**Comment by ngoductuanlhp** [comment]
Hi @aminebdj ,

To get the final predictions from the set of class-agnostic masks and their corresponding CLIP embedding, we follows Mask3D (you can check [these lines](https://github.com/JonasSchult/Mask3D/blob/11bd5ff94477ff7194e9a7c52e9fae54d73ac3b5/trainer/trainer.py#L589-L592) as well as OpenMask3D) to get the `top_k` instances based on the classification scores (based on classification scores (CLIP similarity score in our case)).

To avoid duplicate masks, add the following lines:

```
scores_final, cls_final = torch.max(inst_class_scores, dim=1)
masks_final = instance
```

Best.

**Comment by aminebdj** [comment]
Hey @ngoductuanlhp,

Thank a lot for your reply,

In the case of Mask3D, the Topk is chosen to be equal to the number of queries, which will not result in any duplicate masks; in your case, the TopK is set to 600, which is too high and will result in multiple duplicates masks that will take the top predicted classes from the logits. Since Mask3D uses different score levels, the mAP will not be affected much, but in your case and OpenMask3D case, there are only two points to compute the area under the curve (score 0 (artificial point) and score 1(predicted masks)), so I think it is too sensitive to the number of topk. I believe a fair comparison to OpenMask3D would be to set top-k equal to the number of instances, similar to Mask3D, where they set it equal to the number of queries.

Many thanks,
Mohamed

**Comment by LiangXu123** [comment]
> Hey @ngoductuanlhp, 嘿
> 
> Thank a lot for your reply,非常感谢您的回复，
> 
> In the case of Mask3D, the Topk is chosen to be equal to the number of queries, which will not result in any duplicate masks; in your case, the TopK is set to 600, which is too high and will result in multiple duplicates masks that will take the top predicted classes from the logits.在 Mask3D 的情况下，选择的 Topk 等于查询数，这不会导致任何重复的掩码;在您的例子中，TopK 设置为 600，这太高了，将导致多个重复的掩码，这些掩码将从 logit 中获取排名靠前的预测类。 Since Mask3D uses different score levels, the mAP will not be affected much, but in your case and OpenMask3D case, there are only two points to compute the area under the curve (score 0 (artificial point) and score 1(predicted masks)), so I think it is too sensitive to the number of topk.由于 Mask3D 使用不同的分数水平，mAP 不会受到太大影响，但在你和 OpenMask3D 的情况下，只有两个点可以计算曲线下的面积（得分 0（人工点）和得分 1（预测掩码）），所以我认为它对 topk 的数量太敏感了。 I believe a fair comparison to OpenMask3D would be to set top-k equal to the number of instances, similar to Mask3D, where they set it equal to the number of queries.我相信与 OpenMask3D 的公平比较是将 top-k 设置为等于实例数，类似于 Mask3D，他们将其设置为等于查询数。
> 
> Many thanks, 非常感谢， Mohamed 穆罕默德

same question here, I found that remove the duplicated top k prediction will result in huge performance drop in the final AP, for example the top 600 proposal was reduced to 287, the AP drop from 23 to 18.2, which makes it trickly, because the extra prediction is repeated from the top predictions.

**Comment by shjung13** [comment]
Hi @ngoductuanlhp, I wonder if this same evaluation protocol, _i.e., sampling top-600 / top-300 predictions,_ is applied to other baselines in Table 2 - mainly OpenScene, Sam3D, OVIR-3D, OpenIns3D, and OpenMask3D.
